File: repo/dynamic_trigger.go
package repo

import (
	"qq-krbot/model"
	"time"

	"gorm.io/gorm"
)

type DynamicTriggerModel struct {
	Scene              model.Scene `json:"scene" gorm:"scene"`                             // at pr gr
	ConditionType      string      `json:"conditionType" gorm:"condition_type"`            // equal contains startWith endWith handler
	ConditionValue     string      `json:"conditionValue" gorm:"condition_value"`          //
	TriggerContentType string      `json:"triggerContentType" gorm:"trigger_content_type"` // text image ai api func
	TriggerContent     string      `json:"triggerContent" gorm:"trigger_content"`          // 触发内容
	Sequence           int64       `json:"sequence" gorm:"sequence"`                       // 执行顺序
	Description        string      `json:"description" gorm:"description"`
	gorm.Model
}

type DynamicTriggerRepo struct{}

func NewDynamicTriggerRepo() *DynamicTriggerRepo {
	return &DynamicTriggerRepo{}
}

func (*DynamicTriggerRepo) FindList(model *DynamicTriggerModel) ([]DynamicTriggerModel, error) {
	tx := Sql.Model(&DynamicTriggerModel{})
	if model.Scene != "" {
		tx.Where("scene = ?", model.Scene)
	}
	if model.ConditionType != "" {
		tx.Where("condition_type = ?", model.ConditionType)
	}
	if model.TriggerContentType != "" {
		tx.Where("trigger_content_type = ?", model.TriggerContent)
	}
	if model.TriggerContent != "" {
		tx.Where("trigger_content = ?", model.TriggerContent)
	}
	tx.Order("sequence asc")
	var result []DynamicTriggerModel
	err := tx.Scan(&result).Error
	if err != nil {
		return nil, err
	}
	return result, nil
}

func (*DynamicTriggerRepo) FindOne(id int64) (*DynamicTriggerModel, error) {
	var model DynamicTriggerModel
	err := Sql.Take(&model, "id=?", id).Error
	if err != nil {
		return nil, err
	}
	return &model, nil
}

func (*DynamicTriggerRepo) Delete(id int64) error {
	return Sql.Delete(&DynamicTriggerModel{}, id).Error
}

func (*DynamicTriggerRepo) GetMaxSequence() (int64, error) {
	var maxSequence any
	err := Sql.Model(&DynamicTriggerModel{}).Select("MAX(sequence)").Row().Scan(&maxSequence)
	if err != nil {
		return 0, err
	}
	if maxSequence == nil {
		return 0, nil
	}
	return maxSequence.(int64), nil
}

func (d *DynamicTriggerRepo) Save(model *DynamicTriggerModel) (*DynamicTriggerModel, error) {
	if model.ID != 0 {
		dbModel, err := d.FindOne(int64(model.ID))
		if err != nil {
			return nil, err
		}
		model.UpdatedAt = time.Now()
		model.CreatedAt = dbModel.CreatedAt
	} else {
		model.CreatedAt = time.Now()
		// 使用 GORM 的 Save 方法将数据保存或更新到数据库
		maxSequence, err := d.GetMaxSequence()
		if err != nil {
			return nil, err
		}
		model.Sequence = maxSequence + 1
	}
	if err := Sql.Save(model).Error; err != nil {
		return nil, err // 如果保存或更新失败，返回 nil 和错误信息
	}
	return model, nil // 保存或更新成功，返回保存后的模型和 nil
}

// MoveUp 上移触发器（降低sequence值）
func (d *DynamicTriggerRepo) MoveUp(id int64) error {
	// 获取当前记录
	currentModel, err := d.FindOne(id)
	if err != nil {
		return err
	}

	// 查找前一个记录（sequence更小的）
	var prevModel DynamicTriggerModel
	err = Sql.Where("sequence < ?", currentModel.Sequence).
		Order("sequence DESC").
		First(&prevModel).Error
	if err != nil {
		return err // 已经是第一个了，无法上移
	}

	// 交换sequence值
	currentSeq := currentModel.Sequence
	prevSeq := prevModel.Sequence

	// 开启事务
	tx := Sql.Begin()
	if tx.Error != nil {
		return tx.Error
	}

	// 更新sequence值
	if err := tx.Model(&currentModel).Update("sequence", prevSeq).Error; err != nil {
		tx.Rollback()
		return err
	}

	if err := tx.Model(&prevModel).Update("sequence", currentSeq).Error; err != nil {
		tx.Rollback()
		return err
	}

	return tx.Commit().Error
}

// MoveDown 下移触发器（增加sequence值）
func (d *DynamicTriggerRepo) MoveDown(id int64) error {
	// 获取当前记录
	currentModel, err := d.FindOne(id)
	if err != nil {
		return err
	}

	// 查找后一个记录（sequence更大的）
	var nextModel DynamicTriggerModel
	err = Sql.Where("sequence > ?", currentModel.Sequence).
		Order("sequence ASC").
		First(&nextModel).Error
	if err != nil {
		return err // 已经是最后一个了，无法下移
	}

	// 交换sequence值
	currentSeq := currentModel.Sequence
	nextSeq := nextModel.Sequence

	// 开启事务
	tx := Sql.Begin()
	if tx.Error != nil {
		return tx.Error
	}

	// 更新sequence值
	if err := tx.Model(&currentModel).Update("sequence", nextSeq).Error; err != nil {
		tx.Rollback()
		return err
	}

	if err := tx.Model(&nextModel).Update("sequence", currentSeq).Error; err != nil {
		tx.Rollback()
		return err
	}

	return tx.Commit().Error
}

File: backend/src/repositories/models/dynamic-trigger.ts
import { MessageScene } from '../../types/message.js';
import BaseModel from './base-model.js';

export type ConditionType = 'equals' | 'contains' | 'startsWith' | 'endsWith';

export type TriggerContentType = 'text' | 'api' | 'image' | 'handler';

export type DynamicTriggerData = {
  id: number;
  conditionType: ConditionType;
  conditionValue: string;
  triggerContentType: TriggerContentType;
  triggerContent: string;
  sequence: number;
  scene: MessageScene;
  description: string;
  createdAt: string;
  updatedAt: string;
};

class DynamicTriggerModel extends BaseModel<DynamicTriggerData> {
  table = 'dynamic_trigger_models';

  /**
   * 实现字段映射，定义 TypeScript 类型属性与数据库字段的对应关系
   */
  getFieldMapping(): Record<string, string> {
    return {
      id: 'id',
      conditionType: 'condition_type',
      conditionValue: 'condition_value',
      triggerContentType: 'trigger_content_type',
      triggerContent: 'trigger_content',
      sequence: 'sequence',
      scene: 'scene',
      description: 'description',
      createdAt: 'created_at',
      updatedAt: 'updated_at',
    };
  }

  static async findAll() {
    return await this.query().orderBy('sequence', 'asc').get();
  }
  static delete(id: number[]) {
    return this.query().destroy(id);
  }
  static findOne(id: number) {
    return this.query().where('id', '=', id).first();
  }

  static async moveUp(id: number) {
    const record = await this.query().where('id', '=', id).first();
    if (!record) {
      throw new Error(`Record with id ${id} not found`);
    }

    const currentData = record.getData();

    // 找到上面的记录（sequence 比当前小的最大值）
    const prevRecord = await this.query()
      .where('sequence', '<', currentData.sequence)
      .orderBy('sequence', 'desc')
      .first();

    if (!prevRecord) {
      throw new Error('No record found above the current one');
    }

    const prevData = prevRecord.getData();

    // 交换两个记录的 sequence
    await this.query()
      .where('id', '=', id)
      .update({ sequence: prevData.sequence });
    await this.query()
      .where('id', '=', prevData.id)
      .update({ sequence: currentData.sequence });

    return true;
  }

  static async moveDown(id: number) {
    const record = await this.query().where('id', '=', id).first();
    if (!record) {
      throw new Error(`Record with id ${id} not found`);
    }

    const currentData = record.getData();

    // 找到下面的记录（sequence 比当前大的最小值）
    const nextRecord = await this.query()
      .where('sequence', '>', currentData.sequence)
      .orderBy('sequence', 'asc')
      .first();

    if (!nextRecord) {
      throw new Error('No record found below the current one');
    }

    const nextData = nextRecord.getData();

    // 交换两个记录的 sequence
    await this.query()
      .where('id', '=', id)
      .update({ sequence: nextData.sequence });
    await this.query()
      .where('id', '=', nextData.id)
      .update({ sequence: currentData.sequence });

    return true;
  }
}

export default DynamicTriggerModel;
